crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,404 @@
+                    return cnt;
+                }
+            }
+        }
+    }
+
+    /* buf contains message text; send it to callback */
+    if (ctx->log_cb(ctx->log_func != NULL ? ctx->log_func : "(no func)",
+                    ctx->log_file != NULL ? ctx->log_file : "(no file)",
+                    ctx->log_line, ctx->log_level, buf))
+        return cnt;
+    return 0;
+}
+
+/*
+ * auxiliary function for incrementally reporting texts via the error queue
+ */
+
+void OSSL_CMP_add_error_txt(const char *separator, const char *txt)
+{
+    const char *file;
+    int line;
+    const char *data;
+    int flags;
+    unsigned long err = ERR_peek_last_error();
+
+    if (separator == NULL)
+        separator = "";
+    if (err == 0)
+        ERR_PUT_error(ERR_LIB_CMP, 0, err, "", 0);
+
+#define MAX_DATA_LEN 4096-100 /* workaround for ERR_print_errors_cb() limit */
+    do {
+        const char *curr, *next;
+        int len;
+        char *tmp;
+
+        ERR_peek_last_error_line_data(&file, &line, &data, &flags);
+        if ((flags & ERR_TXT_STRING) == 0) {
+            data = "";
+            separator = "";
+        }
+        len = (int)strlen(data);
+        curr = next = txt;
+        while (*next != '\0'
+                   && len + strlen(separator) + (next - txt) < MAX_DATA_LEN) {
+            curr = next;
+            if (*separator != '\0') {
+                next = strstr(curr, separator);
+                if (next != NULL)
+                    next += strlen(separator);
+                else
+                    next = curr + strlen(curr);
+            } else
+                next = curr + 1;
+        }
+        if (*next != '\0') { /* split error msg if error data gets too long */
+            if (curr != txt) {
+                tmp = OPENSSL_strndup(txt, curr - txt);
+                ERR_add_error_data(2, separator, tmp);
+                OPENSSL_free(tmp);
+            }
+            ERR_PUT_error(ERR_LIB_CMP, 0 /* func */, err, file, line);
+            txt = curr;
+        } else {
+            ERR_add_error_data(2, separator, txt);
+            txt = next;
+        }
+    } while (*txt != '\0');
+}
+
+/* this is similar to ERR_print_errors_cb, but uses the CMP-specific cb type */
+void OSSL_CMP_print_errors_cb(OSSL_cmp_log_cb_t log_fn)
+{
+    unsigned long err;
+    char component[256];
+    char msg[4096];
+    const char *file, *data;
+    int line, flags;
+
+    if (log_fn == NULL) {
+#ifndef OPENSSL_NO_STDIO
+        BIO *bio_err = BIO_new_fp(stderr, BIO_NOCLOSE);
+
+        ERR_print_errors(bio_err);
+        BIO_free(bio_err);
+#endif
+        return;
+    }
+
+    while ((err = ERR_get_error_line_data(&file, &line, &data, &flags)) != 0) {
+        if (!(flags & ERR_TXT_STRING))
+            data = NULL;
+        BIO_snprintf(component, sizeof(component), "OpenSSL:%s",
+                     /* ERR_lib_error_string(err), */
+                     ERR_func_error_string(err));
```

Comment:
The commented out code was a reminder that the lib name could be printed as an alternative or in addition to the function name. Yet since the use of function name (via auto-generated function codes) has just been removed from OpenSSL, I've mentioned this fact in a new comment and replaced the call of `ERR_func_error_string(err)` by `ERR_lib_error_string(err)`.